keps/sig-cli/kubectl-socket-support.md

```diff
@@ -0,0 +1,195 @@
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location
+in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+This proposal adds a new feature that would allow sockets (`unix`, `tcp` and `fd`) as a new protocol for the server to be connected.
+
+## Motivation
+
+Currently several kinds of credentials (passwords, tokens, x509 certificates and etc) are stored in a kubeconfig file and
+persisted on disk. It is not safe and controllable, especially when this kubeconfig file with credentials gets spread to everywhere.
+Moreover, it will become difficult to revoke the access to the apiserver.
+
+Sockets are used as a secured proxy into some infrastructure. For now kubectl does not expose a way to send the request
+over a unix socket. Currently we've only got `http` and `https`. `Docker` client does provide such supports via three
+different types of Socket: `unix`, `tcp`, and `fd`.
+
+It would be great to be able to send requests from kubectl over sockets.
+
+### Goals
+
+* kubectl users can use sockets to connect to the apiserver:
+    * Add an experimental flag `--experimental-socket-addr` for kubectl, much like `-H unix:///var/run/docker.sock -H tcp://192.168.59.106` in `docker` client;
+    * Allow specifying sockets as server addresses in kubeconfig file;
+
+### Non-Goals
+
+* Convey credentials through unix sockets;
+* Adopting [SGX] to secure the memory access;
+
+[SGX]: https://www.intel.com/content/www/us/en/architecture-and-technology/software-guard-extensions.html
+
+## Proposal
+
+### User Stories
+
+Currently `kubectl` supports running proxy on a unix socket.
+
+```bash
+export TOKEN=USER1_TOKEN
+kubectl proxy --unix-socket=/home/user1/.kube/kube.sock --token=${TOKEN}
+```
+
+Issuing commands to any kube apiserver whereby unix socket is the preferred way to access it.
+
+```bash
+kubectl --experimental-socket-addr=/home/user1/.kube/kube.sock get pods
+```
+
+### Implementation Details/Notes/Constraints
+
+A `--experimental-socket-addr` flag will be added to `kubectl`, much like `-H unix:///var/run/docker.sock -H tcp://192.168.59.106` in `docker` client.
```

Comment:
why would we not use the existing `--server` command-line flag which maps to the Server kubeconfig field?

Reply:
+1

Reply:
Seems better.